File: adventofcode/Day8.js
var fs = require('fs');
var readline = require('readline');

var readable = fs.createReadStream(__dirname + '/test.in', {
    encoding: 'utf8'
});
var interface = readline.createInterface({
  input: readable,//process.stdin
  terminal: false
});

var totalDecode = 0;
var totalEncode = 0;
interface.on('line', function(line){
    var decoderCount = function(s){
        var total = 0;
        var i = 0;
        for(var i=0; i < s.length; i++){
            if(s[i] !== '\"'){
                if(s[i] === '\\'){
                    i++;
                    if(s[i] == 'x'){
                        total += 3;
                        i += 2;
                    } else {
                        total += 1
                    }
                }
            } else {
                total++;
            }
        }
        return total;
    };
    var encoderCount = function(s){
        var total = 2;
        for(var i=0; i < s.length; i++)
            if(s[i] === '\"' || s[i] === '\\')
                total++;
        return total;
    };
    totalDecode += decoderCount(line)
    totalEncode += encoderCount(line)
});

interface.on('close', function(){
    console.log(totalDecode, totalEncode)
});

File: adventofcode/Day8.py
def solve():
    totalDecode = 0
    totalEncode = 0
    for line in fin:
        s = line.rstrip()
        totalDecode += decoderCount(s)
        totalEncode += encoderCount(s)
    print(totalDecode, totalEncode)

def decoderCount(s):
    total = 0
    i = 0
    while i < len(s):
        if s[i] != '\"':
            if s[i] == '\\':
                i += 1
                if s[i] == 'x':
                    total += 3
                    i += 2
                else:
                    total += 1
        else:
            total += 1
        i += 1
    return total

def encoderCount(s):
    total = 2
    for c in s:
        if c in ['\"', '\\']:
            total += 1
    return total

fin = open('test.in', 'r')
solve()
fin.close()
